SILModule::isVisibleExternally utility for VarDecls.

atrick opened the pull request.

atrick committed:
SILModule::isVisibleExternally utility for VarDecls.

atrick commented:
@swift-ci test.

atrick commented:
@slavapestov @gottesmm would either of you care to comment on this?

gottesmm commented on lib/SILOptimizer/IPO/DeadFunctionElimination.cpp:
```diff
@@ -373,6 +373,13 @@ class FunctionLivenessComputation {
+  /// This differs from SILModule::isVisibleExternally(VarDecl *) because of
+  /// it's handling of class methods. It returns true for methods whose
+  /// declarations are not directly visible externally, but have been imported
+  /// from another module. This ensures that entries aren't deleted from vtables
```
Great comment!

atrick commented:
@slavapestov unfortunately, making `isVisibleExternally(VarDecl*)` a utility doesn't seem to work in general. Calling getEffectiveAccess on a global Decl crashes (it only works for class properties apparently). The access level is never set by the SIL parser. Is that expected, or a SIL parser bug?

atrick commented:
Sadly, this whole approach is broken because we can't call VarDecl::getEffectiveAccess() from within a SIL pass. I'll close this PR once I figure out the alternative.

atrick commented:
@jrose-apple @slavapestov Should I be able to query a `ValueDecl`'s `AccessLevel` from within a SIL pass? Currently this isn't possible except, apparently, in some special cases. So, should the SIL parser always set the AccessLevel of a Decl, or should SIL passes only relay on the linkage of SIL entities and never query the AccessLevel?

jrose-apple commented:
I would think it's permitted for SIL passes to call getEffectiveAccess. What's the concern here?

jrose-apple commented:
The declarations do need to be type-checked first, so if SIL mode is skipping that for whatever reason then you could end up with this sort of trouble.

atrick commented:
My concern is that calling `getEffectiveAccess` from SIL passes will crash, depending on how/when the decl typechecker is invoked, because the declaration's access level has not been set. It appears to only be an issue with global variables.

atrick commented:
@jrose-apple I'm still debugging this, but it looks like the decl TypeChecker is doing the right thing and setting the access level, even when parsing SIL. However, we end up with alternate, non-uniqued VarDecl objects for these globals that don't have an access level.

atrick committed:
Fix the SIL parser so it doesn't drop global variable decls.

This information was getting lost in SIL printing/parsing.
Some passes rely on it. Regardless of whether passes should rely on it,
it is totally unacceptable for the SIL passes to have subtle differences
in behavior depending on the frontend mode. So, if we don't want passes
to rely on global variable decls, that needs to be enforced by the API
independent of how the frontend is invoked or how SIL is serialized.

atrick commented:
This PR now also fixes the SILParser so that global variable decls are predictable, not silently dropped depending on how the frontend is invoked.

This is a prerequisite for introducing SILGlobalVariable utilities and optimizing global variable access.

Could either @jrose-apple or @eeckstein take a look? It should be a a quick review.

atrick commented:
@swift-ci test.

atrick commented:
@swift-ci test source compatibility.

atrick commented:
Regarding tests, this change will be well tested by the PR for optimizing globals that is blocked on it.

atrick commented:
Unit tests for this PR are now here:
https://github.com/apple/swift/pull/16877/commits/d72c91a242e8ff53b8f78fff5d37c2fcc4fc4c07#diff-871e43cf9543b25afb01e18c18e131e3

eeckstein reviewed (approved):
lgtm. Just make sure it also works for multiple private global variables with the same name

atrick commented:
Before I commit this, I'll add a multi-file SIL parser test case with the same private variable name and different types. I'll be sure that all duplicates are either fileprivate or private and pick the one with the correct type and linkage. If multiple decls have the same name, type, and linkage, I'll add a comment explaining that we just don't care which one we select since they're indistinguishable at that point. This is all fairly sad but I can't think of a better way forward given the design constraints.

atrick commented on include/swift/Parse/ParseSILSupport.h:
```diff
@@ -13,6 +13,7 @@
 #ifndef SWIFT_PARSER_PARSESILSUPPORT_H
 #define SWIFT_PARSER_PARSESILSUPPORT_H
 
+#include "swift/Parse/Parser.h"
```
ParseSILSupport.h refers to a member of the `Parser` class. The #include list was broken before my changes.

atrick commented:
@swift-ci test.

jrose-apple commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -5372,6 +5379,47 @@ bool SILParserTUState::parseDeclSILStage(Parser &P) {
+                                            SILType GlobalType, Parser &P) {
+  std::string GlobalDeclName = Demangle::demangleSymbolAsString(
+      GlobalName.str(),
+      Demangle::DemangleOptions::SimplifiedUIDemangleOptions());
```
Still relevant, please make your own set of demangling options.

atrick replied:
I meant to ask about that. I think these are the right options:
https://github.com/apple/swift/pull/16834/commits/8d9207e2d8468f800ae27c6e832e405378de90de

atrick committed:
Use custom DemangleOptions to lookup global variable identifiers.

atrick commented:
@swift-ci test.

atrick commented:
More test cases will come along in later commits that rely on this feature.

atrick commented:
@swift-ci test linux platform.

atrick commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -5394,9 +5394,15 @@ bool SILParserTUState::parseDeclSILStage(Parser &P) {
+  Demangle::DemangleOptions demangleOpts;
+  demangleOpts.QualifyEntities = false;
+  demangleOpts.ShowPrivateDiscriminators = false;
+  demangleOpts.DisplayEntityTypes = false;
```
@jrose-apple do these demangle options look reasonable?

jrose-apple replied:
Yeah, that looks fine. QualifyEntities does seem to make a number of the other options irrelevant.
